helper/mobile_stream.go

```diff
@@ -0,0 +1,182 @@
+	Data  []byte // Data, the data that has been read
+}
+
+// NewMobileReadResult initialize a MobileReadResult with the correct values.
+// It clones the data to avoid the garbage collector freeing the data too early.
+func NewMobileReadResult(n int, eof bool, data []byte) *MobileReadResult {
+	return &MobileReadResult{N: n, IsEOF: eof, Data: clone(data)}
+}
+
+func clone(src []byte) (dst []byte) {
+	dst = make([]byte, len(src))
+	copy(dst, src)
+	return
+}
+
+// Mobile2GoReader is used to wrap a MobileReader in the mobile app runtime,
+// to be usable in the golang runtime (via gomobile) as a native Reader.
+type Mobile2GoReader struct {
+	reader MobileReader
+}
+
+// NewMobile2GoReader wraps a MobileReader to be usable in the golang runtime (via gomobile).
+func NewMobile2GoReader(reader MobileReader) *Mobile2GoReader {
+	return &Mobile2GoReader{reader}
+}
+
+// Read reads data from the wrapped MobileReader and copies the read data in the provided buffer.
+// It also handles the conversion of EOF to an error.
+func (r *Mobile2GoReader) Read(b []byte) (n int, err error) {
+	result, err := r.reader.Read(len(b))
+	if err != nil {
+		return 0, errors.Wrap(err, "gopenpgp: couldn't read from mobile reader")
+	}
+	n = result.N
+	if n > 0 {
+		copy(b, result.Data[:n])
+	}
+	if result.IsEOF {
+		err = io.EOF
+	}
+	return n, err
+}
+
+// Go2AndroidReader is used to wrap a native golang Reader in the golang runtime,
+// to be usable in the android app runtime (via gomobile).
+type Go2AndroidReader struct {
+	isEOF  bool
+	reader crypto.Reader
+}
+
+// NewGo2AndroidReader wraps a native golang Reader to be usable in the mobile app runtime (via gomobile).
+// It doesn't follow the standard golang Reader behavior, and returns n = -1 on EOF.
+func NewGo2AndroidReader(reader crypto.Reader) *Go2AndroidReader {
+	return &Go2AndroidReader{isEOF: false, reader: reader}
+}
+
+// Read reads bytes into the provided buffer and returns the number of bytes read
+// It doesn't follow the standard golang Reader behavior, and returns n = -1 on EOF.
+func (r *Go2AndroidReader) Read(b []byte) (n int, err error) {
+	if r.isEOF {
+		return -1, nil
+	}
+	n, err = r.reader.Read(b)
+	if errors.Is(err, io.EOF) {
+		if n == 0 {
+			return -1, nil
+		} else {
+			r.isEOF = true
+			return n, nil
+		}
+	}
+	return
+}
+
+// Go2IOSReader is used to wrap a native golang Reader in the golang runtime,
+// to be usable in the iOS app runtime (via gomobile) as a MobileReader.
+type Go2IOSReader struct {
+	reader crypto.Reader
+}
+
+// NewGo2IOSReader wraps a native golang Reader to be usable in the ios app runtime (via gomobile).
+func NewGo2IOSReader(reader crypto.Reader) *Go2IOSReader {
+	return &Go2IOSReader{reader}
+}
+
+// Read reads at most <max> bytes from the wrapped Reader and returns the read data as a MobileReadResult.
+func (r *Go2IOSReader) Read(max int) (result *MobileReadResult, err error) {
```

Comment:
Is `MobileReadResult` only used in iOS? If so, wouldn't it better to be `IOSReadResult`?

Reply:
No, it's also used in android for the MobileReader interface combined with the Mobile2GoReader.
We have wrappers in both directions.

Reply:
OK, then good like this :+1: